File: sdk/python/easynet_sdk/providers/easynet/identity.py
"""EasyNet daemon identity projection adapter."""

from __future__ import annotations

import json
from pathlib import Path
from typing import Mapping

from ...errors import ErrorCode, SDKError
from ...runtime_environment import (
    RuntimeIdentityProjection,
    read_runtime_identity_projection,
)


def read_daemon_runtime_identity_projection(
    credentials_path: str | Path,
) -> RuntimeIdentityProjection:
    """Read EasyNet daemon credentials as a canonical runtime projection."""

    path = Path(credentials_path)
    try:
        return read_runtime_identity_projection(path)
    except SDKError as exc:
        if exc.code != ErrorCode.INVALID_ARGUMENT:
            raise
        raw = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(raw, Mapping):
        raise ValueError("daemon credentials projection must be an object")
    realm = _text(raw, "realm")
    runtime_instance_id = _runtime_instance_id(raw)
    if not realm or not runtime_instance_id:
        raise ValueError("daemon credentials missing runtime identity")
    return RuntimeIdentityProjection(
        realm=realm,
        runtime_instance_id=runtime_instance_id,
        principal=_text(raw, "username"),
        control_plane_endpoint=_text(raw, "hub_endpoint"),
    )


def _runtime_instance_id(raw: Mapping[str, object]) -> str:
    device_id = _text(raw, "device_id")
    node_id = _text(raw, "node_id")
    if device_id and node_id and device_id != node_id:
        raise ValueError("daemon credentials contain conflicting device_id and node_id")
    return device_id or node_id


def _text(raw: Mapping[str, object], key: str) -> str:
    value = raw.get(key)
    if value is None:
        return ""
    return str(value).strip()

File: sdk/go/provider/easynet/identity.go
package easynet

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"

	runtimesdk "easynet.run/cli/sdk/go"
)

// ReadDaemonRuntimeIdentityProjection adapts EasyNet daemon credentials into
// the canonical runtime identity projection.
//
// EasyNet credentials are product-owned and expose a device identity fact. The
// canonical SDK root does not accept product fields as runtime identity; this
// provider adapter is the only place that translates device_id into
// RuntimeInstanceID.
func ReadDaemonRuntimeIdentityProjection(ctx context.Context, credentialsPath string) (runtimesdk.RuntimeIdentityProjection, error) {
	projection, err := runtimesdk.ReadRuntimeIdentityProjection(ctx, credentialsPath, "")
	if err == nil {
		return projection, nil
	}
	if !runtimesdk.IsCode(err, runtimesdk.ErrInvalidArgument) {
		return runtimesdk.RuntimeIdentityProjection{}, err
	}
	raw, readErr := os.ReadFile(strings.TrimSpace(credentialsPath))
	if readErr != nil {
		return runtimesdk.RuntimeIdentityProjection{}, readErr
	}
	var decoded map[string]any
	if unmarshalErr := json.Unmarshal(raw, &decoded); unmarshalErr != nil {
		return runtimesdk.RuntimeIdentityProjection{}, unmarshalErr
	}
	realm := providerIdentityString(decoded, "realm")
	runtimeInstanceID, idErr := providerRuntimeInstanceID(decoded)
	if idErr != nil {
		return runtimesdk.RuntimeIdentityProjection{}, idErr
	}
	if realm == "" || runtimeInstanceID == "" {
		return runtimesdk.RuntimeIdentityProjection{}, fmt.Errorf("daemon credentials missing runtime identity")
	}
	return runtimesdk.RuntimeIdentityProjection{
		Realm:                realm,
		RuntimeInstanceID:    runtimeInstanceID,
		Principal:            providerIdentityString(decoded, "username"),
		ControlPlaneEndpoint: providerIdentityString(decoded, "hub_endpoint"),
	}, nil
}

func providerRuntimeInstanceID(decoded map[string]any) (string, error) {
	if retired := providerIdentityString(decoded, "node_id"); retired != "" {
		return "", fmt.Errorf("daemon credentials contain retired node_id identity alias")
	}
	deviceID := providerIdentityString(decoded, "device_id")
	return deviceID, nil
}

func providerIdentityString(raw map[string]any, key string) string {
	value, ok := raw[key]
	if !ok || value == nil {
		return ""
	}
	if text, ok := value.(string); ok {
		return strings.TrimSpace(text)
	}
	return strings.TrimSpace(fmt.Sprint(value))
}
